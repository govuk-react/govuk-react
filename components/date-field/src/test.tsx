import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, mount } from 'enzyme';

import { DateField } from '.';

// TODO: all snapshot tests in this file should be moved to fixtures and included in stories

describe('DateField', () => {
  const example = 'example';
  const wrapper = <DateField errorText={example}>{example}</DateField>;
  let props;

  beforeEach(() => {
    props = {
      children: example,
    };
  });

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <DateField errorText={example} hintText={example}>
        {example}
      </DateField>,
      div
    );
  });

  it('should render an input type="text"', () => {
    const output = shallow(wrapper);
    expect(output.find('input[type="text"]')).toBeTruthy();
  });

  it('should render one input with name attributes if only a day input name is passed', () => {
    const output = mount(
      <DateField inputNames={{ day: 'dayInputName' }} errorText={example}>
        {example}
      </DateField>
    );
    expect(output.find('input[name]')).toHaveLength(1);
  });

  it('should render no defaultValue attributes if defaultValues are not passed', () => {
    const output = mount(wrapper);
    expect(output.find('input[defaultValue]')).toHaveLength(0);
  });

  it('should render two defaultValue attributes if two defaultValues are passed', () => {
    const output = mount(
      <DateField defaultValues={{ day: '1', month: '2' }} errorText={example}>
        {example}
      </DateField>
    );
    expect(output.find('input[defaultValue]')).toHaveLength(2);
  });

  it('should render a label', () => {
    const output = shallow(wrapper);
    expect(output.find('label')).toBeTruthy();
  });

  it('matches wrappersnapshot', () => {
    expect(mount(wrapper)).toMatchSnapshot('wrapper mount');
  });

  it('should support setting value', () => {
    const output = mount(<DateField input={{ value: { day: '1', month: '2', year: '3' } }}>Date</DateField>);
    expect(output).toMatchSnapshot('value 1 2 3');
    expect(output.find('input').first().props().value).toBe('1');
  });

  it('should support null value', () => {
    const output = mount(<DateField input={{ value: null }}>Date</DateField>);
    expect(output).toMatchSnapshot('null value');
  });

  it('passes `props` to the rendered label', () => {
    expect(mount(<DateField errorText={example}>{example}</DateField>).find('input')).toHaveLength(3);
    expect(mount(<DateField hintText={example}>{example}</DateField>).find('div')).toHaveLength(2);
  });

  it('passes `props.children` to the rendered `wrapper` as `children`', () => {
    expect(wrapper.props.children).toBe(props.children);
  });

  it('calls onBlur', () => {
    const spy = jest.fn();

    mount(<DateField onBlur={spy}>Date</DateField>)
      .find('input')
      .first()
      .simulate('blur');
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it('calls onFocus', () => {
    const spy = jest.fn();

    mount(<DateField onFocus={spy}>Date</DateField>)
      .find('input')
      .first()
      .simulate('focus');
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it('does not call onFocus when moving between fields ', () => {
    const spy = jest.fn();

    const inst = mount(
      <DateField
        input={{
          onFocus: (...rest) => {
            spy(...rest);
          },
        }}
      >
        Date
      </DateField>
    );
    const input1 = inst.find('input').first();
    const input2 = inst.find('input').at(1);

    // Focus first input
    input1.simulate('focus', {});
    // Then simulate tab to second input, first by blurring first input
    input1.simulate('blur', {
      relatedTarget: input2.instance(), // relatedTarget for blur is what will next receive focus
    });
    // Then focus the new input
    input2.simulate('focus', {
      relatedTarget: input1.instance(), // relatedTarget for focus is what has lost focus
    });
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it('calls onChange', () => {
    const spy = jest.fn();

    mount(<DateField onChange={spy}>Date</DateField>)
      .find('input')
      .first()
      .simulate('change');
    expect(spy).toHaveBeenCalledTimes(1);
  });

  // TODO: works controlled and uncontrolled
  // TODO: onBlur, onChange, onFocus work as expected when tabbing between fields,
  //       both controlled and uncontrolled
  // TODO: defaultValues work controlled and uncontrolled
});
