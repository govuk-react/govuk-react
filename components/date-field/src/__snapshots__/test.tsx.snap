// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateField matches wrappersnapshot: wrapper mount 1`] = `
.c1 {
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 400;
  font-size: 16px;
  line-height: 1.25;
  color: #0b0c0c;
  display: block;
  clear: none;
  padding-bottom: 2px;
  margin-bottom: 0;
}

.c2 {
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 700;
  font-size: 16px;
  line-height: 1.25;
  display: block;
  margin-bottom: 15px;
  clear: both;
  color: #d4351c;
}

.c6 {
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 400;
  font-size: 16px;
  line-height: 1.25;
  box-sizing: border-box;
  width: 100%;
  height: 40px;
  margin-top: 0;
  padding: 5px;
  border: 2px solid #0b0c0c;
  border-radius: 0;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border: 4px solid #d4351c;
  margin-bottom: 0;
}

.c6:focus {
  outline: 3px solid #ffdd00;
  outline-offset: 0;
}

.c6::-webkit-outer-spin-button,
.c6::-webkit-inner-spin-button {
  margin: 0;
  -webkit-appearance: none;
}

.c6[type="number"] {
  -moz-appearance: textfield;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  box-sizing: border-box;
  margin-bottom: 0;
}

.c4:after {
  content: '';
  display: table;
  clear: both;
}

.c5 {
  margin-right: 20px;
  margin-bottom: 0;
  width: 50px;
}

.c7 {
  margin-right: 20px;
  margin-bottom: 0;
  width: 70px;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  border-left: 4px solid #d4351c;
  margin-right: 15px;
  padding-left: 10px;
  margin-bottom: 20px;
}

@media print {
  .c1 {
    font-size: 14px;
    line-height: 1.15;
  }
}

@media only screen and (min-width:641px) {
  .c1 {
    font-size: 19px;
    line-height: 1.3157894736842106;
  }
}

@media print {
  .c1 {
    color: #000;
  }
}

@media only screen and (min-width:641px) {
  .c1 {
    margin-bottom: 0;
  }
}

@media print {
  .c2 {
    font-size: 14px;
    line-height: 1.15;
  }
}

@media only screen and (min-width:641px) {
  .c2 {
    font-size: 19px;
    line-height: 1.3157894736842106;
  }
}

@media print {
  .c6 {
    font-size: 14px;
    line-height: 1.15;
  }
}

@media only screen and (min-width:641px) {
  .c6 {
    font-size: 19px;
    line-height: 1.3157894736842106;
  }
}

@media only screen and (min-width:641px) {
  .c6 {
    margin-bottom: 0;
  }
}

@media only screen and (min-width:641px) {
  .c4 {
    margin-bottom: 0;
  }
}

@media only screen and (min-width:641px) {
  .c0 {
    margin-bottom: 30px;
  }
}

<DateField
  defaultValues={
    Object {
      "day": undefined,
      "month": undefined,
      "year": undefined,
    }
  }
  errorText="example"
  inputNames={
    Object {
      "day": undefined,
      "month": undefined,
      "year": undefined,
    }
  }
>
  <src__StyledContainer
    errorText="example"
  >
    <div
      className="c0"
    >
      <LabelText>
        <span
          className="c1"
        >
          example
        </span>
      </LabelText>
      <ErrorText>
        <span
          className="c2"
        >
          example
        </span>
      </ErrorText>
      <MultiInputInput
        defaultValues={
          Object {
            "day": undefined,
            "month": undefined,
            "year": undefined,
          }
        }
        error={true}
        names={
          Object {
            "day": undefined,
            "month": undefined,
            "year": undefined,
          }
        }
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        value={Object {}}
      >
        <Input
          defaultValues={
            Object {
              "day": undefined,
              "month": undefined,
              "year": undefined,
            }
          }
          error={true}
          labels={
            Object {
              "day": "Day",
              "month": "Month",
              "year": "Year",
            }
          }
          names={
            Object {
              "day": undefined,
              "month": undefined,
              "year": undefined,
            }
          }
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          refs={[Function]}
          value={Object {}}
        >
          <input__StyledList>
            <div
              className="c3"
            >
              <input__StyledLabel
                error={false}
                year={false}
              >
                <label
                  className="c4 c5"
                >
                  <LabelText>
                    <span
                      className="c1"
                    >
                      Day
                    </span>
                  </LabelText>
                  <Input
                    error={true}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    type="text"
                  >
                    <input
                      className="c6"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      type="text"
                    />
                  </Input>
                </label>
              </input__StyledLabel>
              <input__StyledLabel
                error={false}
                year={false}
              >
                <label
                  className="c4 c5"
                >
                  <LabelText>
                    <span
                      className="c1"
                    >
                      Month
                    </span>
                  </LabelText>
                  <Input
                    error={true}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    type="text"
                  >
                    <input
                      className="c6"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      type="text"
                    />
                  </Input>
                </label>
              </input__StyledLabel>
              <input__StyledLabel
                error={false}
                year={true}
              >
                <label
                  className="c4 c7"
                >
                  <LabelText>
                    <span
                      className="c1"
                    >
                      Year
                    </span>
                  </LabelText>
                  <Input
                    error={true}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    type="text"
                  >
                    <input
                      className="c6"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      type="text"
                    />
                  </Input>
                </label>
              </input__StyledLabel>
            </div>
          </input__StyledList>
        </Input>
      </MultiInputInput>
    </div>
  </src__StyledContainer>
</DateField>
`;

exports[`DateField should support null value: null value 1`] = `
.c1 {
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 400;
  font-size: 16px;
  line-height: 1.25;
  color: #0b0c0c;
  display: block;
  clear: none;
  padding-bottom: 2px;
  margin-bottom: 0;
}

.c5 {
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 400;
  font-size: 16px;
  line-height: 1.25;
  box-sizing: border-box;
  width: 100%;
  height: 40px;
  margin-top: 0;
  padding: 5px;
  border: 2px solid #0b0c0c;
  border-radius: 0;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  margin-bottom: 0;
}

.c5:focus {
  outline: 3px solid #ffdd00;
  outline-offset: 0;
}

.c5::-webkit-outer-spin-button,
.c5::-webkit-inner-spin-button {
  margin: 0;
  -webkit-appearance: none;
}

.c5[type="number"] {
  -moz-appearance: textfield;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  box-sizing: border-box;
  margin-bottom: 0;
}

.c3:after {
  content: '';
  display: table;
  clear: both;
}

.c4 {
  margin-right: 20px;
  margin-bottom: 0;
  width: 50px;
}

.c6 {
  margin-right: 20px;
  margin-bottom: 0;
  width: 70px;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-bottom: 20px;
}

@media print {
  .c1 {
    font-size: 14px;
    line-height: 1.15;
  }
}

@media only screen and (min-width:641px) {
  .c1 {
    font-size: 19px;
    line-height: 1.3157894736842106;
  }
}

@media print {
  .c1 {
    color: #000;
  }
}

@media only screen and (min-width:641px) {
  .c1 {
    margin-bottom: 0;
  }
}

@media print {
  .c5 {
    font-size: 14px;
    line-height: 1.15;
  }
}

@media only screen and (min-width:641px) {
  .c5 {
    font-size: 19px;
    line-height: 1.3157894736842106;
  }
}

@media only screen and (min-width:641px) {
  .c5 {
    margin-bottom: 0;
  }
}

@media only screen and (min-width:641px) {
  .c3 {
    margin-bottom: 0;
  }
}

@media only screen and (min-width:641px) {
  .c0 {
    margin-bottom: 30px;
  }
}

<DateField
  defaultValues={
    Object {
      "day": undefined,
      "month": undefined,
      "year": undefined,
    }
  }
  input={
    Object {
      "value": null,
    }
  }
  inputNames={
    Object {
      "day": undefined,
      "month": undefined,
      "year": undefined,
    }
  }
>
  <src__StyledContainer>
    <div
      className="c0"
    >
      <LabelText>
        <span
          className="c1"
        >
          Date
        </span>
      </LabelText>
      <MultiInputInput
        defaultValues={
          Object {
            "day": undefined,
            "month": undefined,
            "year": undefined,
          }
        }
        error={false}
        names={
          Object {
            "day": undefined,
            "month": undefined,
            "year": undefined,
          }
        }
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        value={null}
      >
        <Input
          defaultValues={
            Object {
              "day": undefined,
              "month": undefined,
              "year": undefined,
            }
          }
          error={false}
          labels={
            Object {
              "day": "Day",
              "month": "Month",
              "year": "Year",
            }
          }
          names={
            Object {
              "day": undefined,
              "month": undefined,
              "year": undefined,
            }
          }
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          refs={[Function]}
          value={null}
        >
          <input__StyledList>
            <div
              className="c2"
            >
              <input__StyledLabel
                error={false}
                year={false}
              >
                <label
                  className="c3 c4"
                >
                  <LabelText>
                    <span
                      className="c1"
                    >
                      Day
                    </span>
                  </LabelText>
                  <Input
                    error={false}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    type="text"
                  >
                    <input
                      className="c5"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      type="text"
                    />
                  </Input>
                </label>
              </input__StyledLabel>
              <input__StyledLabel
                error={false}
                year={false}
              >
                <label
                  className="c3 c4"
                >
                  <LabelText>
                    <span
                      className="c1"
                    >
                      Month
                    </span>
                  </LabelText>
                  <Input
                    error={false}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    type="text"
                  >
                    <input
                      className="c5"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      type="text"
                    />
                  </Input>
                </label>
              </input__StyledLabel>
              <input__StyledLabel
                error={false}
                year={true}
              >
                <label
                  className="c3 c6"
                >
                  <LabelText>
                    <span
                      className="c1"
                    >
                      Year
                    </span>
                  </LabelText>
                  <Input
                    error={false}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    type="text"
                  >
                    <input
                      className="c5"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      type="text"
                    />
                  </Input>
                </label>
              </input__StyledLabel>
            </div>
          </input__StyledList>
        </Input>
      </MultiInputInput>
    </div>
  </src__StyledContainer>
</DateField>
`;

exports[`DateField should support setting value: value 1 2 3 1`] = `
.c1 {
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 400;
  font-size: 16px;
  line-height: 1.25;
  color: #0b0c0c;
  display: block;
  clear: none;
  padding-bottom: 2px;
  margin-bottom: 0;
}

.c5 {
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 400;
  font-size: 16px;
  line-height: 1.25;
  box-sizing: border-box;
  width: 100%;
  height: 40px;
  margin-top: 0;
  padding: 5px;
  border: 2px solid #0b0c0c;
  border-radius: 0;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  margin-bottom: 0;
}

.c5:focus {
  outline: 3px solid #ffdd00;
  outline-offset: 0;
}

.c5::-webkit-outer-spin-button,
.c5::-webkit-inner-spin-button {
  margin: 0;
  -webkit-appearance: none;
}

.c5[type="number"] {
  -moz-appearance: textfield;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  box-sizing: border-box;
  margin-bottom: 0;
}

.c3:after {
  content: '';
  display: table;
  clear: both;
}

.c4 {
  margin-right: 20px;
  margin-bottom: 0;
  width: 50px;
}

.c6 {
  margin-right: 20px;
  margin-bottom: 0;
  width: 70px;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-bottom: 20px;
}

@media print {
  .c1 {
    font-size: 14px;
    line-height: 1.15;
  }
}

@media only screen and (min-width:641px) {
  .c1 {
    font-size: 19px;
    line-height: 1.3157894736842106;
  }
}

@media print {
  .c1 {
    color: #000;
  }
}

@media only screen and (min-width:641px) {
  .c1 {
    margin-bottom: 0;
  }
}

@media print {
  .c5 {
    font-size: 14px;
    line-height: 1.15;
  }
}

@media only screen and (min-width:641px) {
  .c5 {
    font-size: 19px;
    line-height: 1.3157894736842106;
  }
}

@media only screen and (min-width:641px) {
  .c5 {
    margin-bottom: 0;
  }
}

@media only screen and (min-width:641px) {
  .c3 {
    margin-bottom: 0;
  }
}

@media only screen and (min-width:641px) {
  .c0 {
    margin-bottom: 30px;
  }
}

<DateField
  defaultValues={
    Object {
      "day": undefined,
      "month": undefined,
      "year": undefined,
    }
  }
  input={
    Object {
      "value": Object {
        "day": "1",
        "month": "2",
        "year": "3",
      },
    }
  }
  inputNames={
    Object {
      "day": undefined,
      "month": undefined,
      "year": undefined,
    }
  }
>
  <src__StyledContainer>
    <div
      className="c0"
    >
      <LabelText>
        <span
          className="c1"
        >
          Date
        </span>
      </LabelText>
      <MultiInputInput
        defaultValues={
          Object {
            "day": undefined,
            "month": undefined,
            "year": undefined,
          }
        }
        error={false}
        names={
          Object {
            "day": undefined,
            "month": undefined,
            "year": undefined,
          }
        }
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        value={
          Object {
            "day": "1",
            "month": "2",
            "year": "3",
          }
        }
      >
        <Input
          defaultValues={
            Object {
              "day": undefined,
              "month": undefined,
              "year": undefined,
            }
          }
          error={false}
          labels={
            Object {
              "day": "Day",
              "month": "Month",
              "year": "Year",
            }
          }
          names={
            Object {
              "day": undefined,
              "month": undefined,
              "year": undefined,
            }
          }
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          refs={[Function]}
          value={
            Object {
              "day": "1",
              "month": "2",
              "year": "3",
            }
          }
        >
          <input__StyledList>
            <div
              className="c2"
            >
              <input__StyledLabel
                error={false}
                year={false}
              >
                <label
                  className="c3 c4"
                >
                  <LabelText>
                    <span
                      className="c1"
                    >
                      Day
                    </span>
                  </LabelText>
                  <Input
                    error={false}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    type="text"
                    value="1"
                  >
                    <input
                      className="c5"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      type="text"
                      value="1"
                    />
                  </Input>
                </label>
              </input__StyledLabel>
              <input__StyledLabel
                error={false}
                year={false}
              >
                <label
                  className="c3 c4"
                >
                  <LabelText>
                    <span
                      className="c1"
                    >
                      Month
                    </span>
                  </LabelText>
                  <Input
                    error={false}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    type="text"
                    value="2"
                  >
                    <input
                      className="c5"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      type="text"
                      value="2"
                    />
                  </Input>
                </label>
              </input__StyledLabel>
              <input__StyledLabel
                error={false}
                year={true}
              >
                <label
                  className="c3 c6"
                >
                  <LabelText>
                    <span
                      className="c1"
                    >
                      Year
                    </span>
                  </LabelText>
                  <Input
                    error={false}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    type="text"
                    value="3"
                  >
                    <input
                      className="c5"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      type="text"
                      value="3"
                    />
                  </Input>
                </label>
              </input__StyledLabel>
            </div>
          </input__StyledList>
        </Input>
      </MultiInputInput>
    </div>
  </src__StyledContainer>
</DateField>
`;
